### PR TITLE
Remove unused headers from gsl/pointers

### PR DESCRIPTION
**Pull request by knst**
- `forward` is already declared in utility, no need to include `<algorithm>` which is relatively heavy
 - `hash` is already declared in memory, no need to bring brand-new header `<system_error>` for hash only.

On my machine with my customized fork of GSL it improves a compilation time of gsl/pointers drastically.

See benchmark (best time over 5 runs):

 - baseline
```
$ time g++ -o a.out gsl/pointers.h -O2 -g -I.
real    0m0,572s
user    0m0,461s
sys     0m0,108s
```

 - removed algorithm:
```
real    0m0,505s
user    0m0,398s
sys     0m0,107s
```

 - without algorithm and system_error:
```
real    0m0,332s
user    0m0,265s
sys     0m0,067s
```

Considering using `gsl/pointers` all over codebase it speeds up compilation time of any project that uses gsl/pointers.

### CODE REVIEW

File: include/gsl/pointers [hunk 1]
```diff
@@ -19,12 +19,10 @@
...
 #include "assert" // for Ensures, Expects
 
-#include <algorithm>    // for forward
 #include <cstddef>      // for ptrdiff_t, nullptr_t, size_t
-#include <memory>       // for shared_ptr, unique_ptr
-#include <system_error> // for hash
+#include <memory>       // for shared_ptr, unique_ptr, for hash
 #include <type_traits>  // for enable_if_t, is_convertible, is_assignable
-#include <utility>      // for declval
+#include <utility>      // for declval, for forward
```
Comment (beinhaerter) [inline include/gsl/pointers:25]:
```suggestion
#include <utility>      // for declval, forward
```

File: include/gsl/pointers [hunk 2]
```diff
@@ -19,12 +19,10 @@
 
 #include "assert" // for Ensures, Expects
 
-#include <algorithm>    // for forward
 #include <cstddef>      // for ptrdiff_t, nullptr_t, size_t
-#include <memory>       // for shared_ptr, unique_ptr
-#include <system_error> // for hash
+#include <memory>       // for shared_ptr, unique_ptr, for hash
```
Comment (beinhaerter) [inline include/gsl/pointers:23]:
```suggestion
#include <memory>       // for shared_ptr, unique_ptr, hash
```

**Review by beinhaerter — APPROVED** [review]

**Comment by knst** [comment]
hi! Is there any chances too see a progress here? Can it get merged to the main?

it seems useful changes in my opinion.

**Review by gdr-at-ms — APPROVED** [review]

**Review by d-winsor — APPROVED** [review]

**Comment by d-winsor** [comment]
@knst There appears to be a problem running checks on this PR. Can you push a dummy change as shown below to retrigger the tests.

```commit -m "retrigger checks" --allow-empty```

**Comment by knst** [comment]
@d-winsor 
>  3 workflows awaiting approval 

> problem running checks on this PR

I force pushed a branch, seems as I need approval again

**Comment by PastaPastaPasta** [comment]
> @knst There appears to be a problem running checks on this PR. Can you push a dummy change as shown below to retrigger the tests.
> 
> `commit -m "retrigger checks" --allow-empty`

Hi @d-winsor, it appears the CI is still not triggering? Please advise.

Thanks!

**Comment by PastaPastaPasta** [comment]
Hey all, 

this has been approved for a while? CI failures seem unrelated? Please advise, it'd be nice to get some movement again

**Comment by carsonRadtke** [comment]
@knst Thanks for the contribution!

Apologies it took so long to get the CI issues sorted out.

**Comment by knst** [comment]
@carsonRadtke thanks a lot for getting it merged!

**Pull request merged**